### DISCUSSION TITLE
01_pytorch_workflow_exercises question

### DISCUSSION THREAD

**Discussion by sharkweekshane** [mrdbourke/pytorch-deep-learning#260] [Q&A]
Hi folks,
Been loving the course so far (Cue fire animation)

I'm working through the second set of exercises (01) and I keep getting an error when I try to plot the predictions

# Plot the predictions (these may need to be on a specific device)
`plot_predictions(predictions = y_preds.cpu())`

I get error: `RuntimeError: Can't call numpy() on Tensor that requires grad. Use tensor.detach().numpy() instead.`

I see a similar question was asked about this, but I can't wrap my mind around it. Can someone help? 
  
Edit: I also tried `plot_predictions(predictions = y_preds.cpu().detach().numpy())` as suggested by the error, and I got another error that I also don't understand

**Comment by DavS123** [comment]
Hi,

y_preds.cpu().detach().numpy() should we replaces by y_preds.detach().numpy().cpu()

should do the trick.

best wishes

**Reply by sharkweekshane** [reply]
Hi Dav,
Thanks for the reply. I tried doing it your way and ended up with another error (sigh) 

input: `plot_predictions(predictions = y_preds.detach().numpy().cpu())`
output: `AttributeError: 'numpy.ndarray' object has no attribute 'cpu'` 

I'm still a novice coder, so I'm not sure how to tackle this error. My inclination tells me that I should try converting the tensor to an array before I try ploting.

**Reply by DavS123** [reply]
Hi 
here is a good reference.

https://www.tutorialspoint.com/how-to-convert-a-pytorch-tensor-with-gradient-to-a-numpy-array

Apprently y_preds.detach().numpy() does the trick.

best wishes

**Comment by mrdbourke** [comment]
Hi @sharkweekshane,

I've faced that error plenty of times too. And it can be a bit hard to troubleshoot.

I just went through the solutions notebook and it seems to work fine end-to-end: https://github.com/mrdbourke/pytorch-deep-learning/blob/main/extras/solutions/01_pytorch_workflow_exercise_solutions.ipynb 

I'd say a potential issue you may be facing:
* Data already on `cpu()` (this happens if you haven't set the device to GPU)/sent to GPU

You can check what device your PyTorch tensor is on with `tensor.device`:

```python
y_preds.device
```

Out:
```
device(type='cuda', index=0)
```

```python
y_preds.cpu().device
```

Out:
```
device(type='cpu')
```

However, `tensor.device` will error *if* your tensor is already NumPy.

As for this line:

```python
tensor.cpu().detach().numpy()
```

This is saying "send `tensor` to the `cpu()`, `detach()` the gradient tracking (this is automatically set for learning in PyTorch) and then turn the PyTorch tesnor to `numpy()` datatype".

Hopefully this clears things up!
